### ISSUE TITLE
ipywidgets 5.0 breaks AMD/require.js

### ISSUE THREAD

**Issue by jhpedemonte** [ipython/ipywidgets#561]
Offshoot from https://github.com/jupyter-incubator/dashboards_server/issues/194.

If building *jupyter-js-widgets* as an AMD module to be used by require.js, loading of widgets is broken. See issue referenced above for history. Basically, [this change](https://github.com/ipython/ipywidgets/commit/fa33fc14b3b0b119e0a8e4d7c5aa72806cf226c3) in utils.js seems to introduce a timing issue, such that loading modules aren't handled at the correct time and model changes aren't correctly handled.

I [rewrote the web3 example](https://github.com/jhpedemonte/ipywidgets/tree/web3-requirejs) to build *jupyter-js-widgets* as an AMD module. When reloading the page, the widgets fail to render roughly 50% of the time. Compare to web3 on master, where they render 100% of the time.

I spent some time debugging https://github.com/jupyter-incubator/dashboards_server/issues/194 but was not able to find out where the timing issue lay.

cc @SylvainCorlay @parente

**Comment by jhpedemonte** [comment]
There is a deeper nefarious timing issue going on within ipywidgets. We've run into another issue (https://github.com/jupyter-incubator/dashboards_server/issues/240) where ipywidgets fail to load correctly.  And @SylvainCorlay has seen similar issues.

I was able to reproduce this issue with minimal changes to the *web3* example. See this commit: https://github.com/jhpedemonte/ipywidgets/commit/73320ab779945969a581ff3375a26ec3318132ad -- reloading *web3* fails to properly render the widgets roughly 50% of the time.

---

Here's what we know:

1. There seems to be a timing issue deep in ipywidgets. In the simple case, where require.js isn't used, `utils.loadClass()` directly loads `jupyter-js-widgets` code **synchronously**. In certain situations where require.js is used and `jupyter-js-widgets` has already been required (i.e. by 3rd party widgets), then it goes down the path of calling `window.require` (yes, even for `jupyter-js-widgets`) which means that it is now loaded **asynchronously**. This seems to result in a timing issue where widgets aren't rendered correctly.
2. There is nothing inherently wrong with the `loadClass()` code -- it correctly uses Promises. So the timing issue must be elsewhere, where it expects/assumes that `loadClass` will return synchronously.
3. That being said, this line:

    ```js
    if (module_name !== 'jupyter-js-widgets' || window.requirejs.defined('jupyter-js-widgets')) {
    ```
is strange and doesn't make sense to me. This means that when `jupyter-js-widgets` is defined in require.js, it will always call `window.require` to load it, when going down the `else` path works just as well (and makes more sense in this case).  Why was that `or` clause added to this code? What is the purpose?  Removing the 2nd part of the `or` would alleviate this problem, although it wouldn't be a fix for the timing issue.

**Comment by jhpedemonte** [comment]
Added some logging around comm messages:

### Success case:
```
Starting WebSocket: ws://192.168.99.100:9500/api/kernels/d71f358d-8bd9-4469-b133-af5dd0032954
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "16c4a8d3149147b4b921a53823c326cb"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 16c4a8d3149147b4b921a53823c326cb
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "80c64f9ff45b438c987b8d86c78293ac"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 80c64f9ff45b438c987b8d86c78293ac
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "d80358493ec94c348ca736f528493336"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id d80358493ec94c348ca736f528493336
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "a9d0b332e98145fcb34d22f8b0426793"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id a9d0b332e98145fcb34d22f8b0426793
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "bbe9c69c24da4200bae0e4ecf6871de9"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id bbe9c69c24da4200bae0e4ecf6871de9
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "fa1d0a65945f465292719bf7a4a7c334"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id fa1d0a65945f465292719bf7a4a7c334
comm_msg Object {data: Object, comm_id: "bbe9c69c24da4200bae0e4ecf6871de9"}
-- handling comm_msg
comm_msg Object {data: Object, comm_id: "bbe9c69c24da4200bae0e4ecf6871de9"}
-- handling comm_msg
comm_msg Object {data: Object, comm_id: "fa1d0a65945f465292719bf7a4a7c334"}
-- handling comm_msg
-- sync -- require of jupyter-js-widgets
-- sync -- require of jupyter-js-widgets
-- sync -- require of jupyter-js-widgets
-- sync -- require of jupyter-js-widgets
```

### Failure case:
```
Starting WebSocket: ws://192.168.99.100:9500/api/kernels/db8f9e6b-0d26-4dd5-a7ab-96b5308c669d
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "03e277ee19a84444894e3e28233d082d"}
-- async -- window.require of jupyter-js-widgets
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "0df5a65659ba4c9a919205aa42d09686"}
-- async -- window.require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 03e277ee19a84444894e3e28233d082d
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "3ddd27ee2cd54be19cd5a7ce58f83119"}
-- async -- window.require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 0df5a65659ba4c9a919205aa42d09686
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "096e0ab299ae4dcaad82b35f34ec3ba7"}
-- async -- window.require of jupyter-js-widgets
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "7cb420941ee24271a2e840eac496b2a9"}
-- async -- window.require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 3ddd27ee2cd54be19cd5a7ce58f83119
comm_open Object {target_module: null, data: Object, target_name: "jupyter.widget", comm_id: "f6c31741c06e4c5f9891ab2fb7c4c1a7"}
-- async -- window.require of jupyter-js-widgets
comm_msg Object {data: Object, comm_id: "7cb420941ee24271a2e840eac496b2a9"}
new_model:  MODEL CREATED for comm_id 096e0ab299ae4dcaad82b35f34ec3ba7
comm_msg Object {data: Object, comm_id: "7cb420941ee24271a2e840eac496b2a9"}
new_model:  MODEL CREATED for comm_id 7cb420941ee24271a2e840eac496b2a9
comm_msg Object {data: Object, comm_id: "f6c31741c06e4c5f9891ab2fb7c4c1a7"}
new_model:  MODEL CREATED for comm_id f6c31741c06e4c5f9891ab2fb7c4c1a7
```

Notice in the failure case that `comm_msg` arrive for some IDs before the model has been created. Therefore, those messages aren't handled (notice no instances of `handling comm_msg`).

We probably need to use promises from `comm_open` in order to chain incoming `comm_msg` (for same ID). That way, the `comm_open` will handle finish being handled (i.e. model created) before any `comm_msg`s are handled.

**Comment by SylvainCorlay** [comment]
Thanks for looking into this! I agree about the promises chaining although
the thing you were pointing to in loadClass looks ok to me.
On May 19, 2016 12:10 PM, "Javier Pedemonte" <notifications@github.com>
wrote:

Added some logging around comm messages:
Success case:

Starting WebSocket:
ws://192.168.99.100:9500/api/kernels/d71f358d-8bd9-4469-b133-af5dd0032954
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "16c4a8d3149147b4b921a53823c326cb"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 16c4a8d3149147b4b921a53823c326cb
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "80c64f9ff45b438c987b8d86c78293ac"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 80c64f9ff45b438c987b8d86c78293ac
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "d80358493ec94c348ca736f528493336"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id d80358493ec94c348ca736f528493336
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "a9d0b332e98145fcb34d22f8b0426793"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id a9d0b332e98145fcb34d22f8b0426793
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "bbe9c69c24da4200bae0e4ecf6871de9"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id bbe9c69c24da4200bae0e4ecf6871de9
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "fa1d0a65945f465292719bf7a4a7c334"}
-- sync -- require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id fa1d0a65945f465292719bf7a4a7c334
comm_msg Object {data: Object, comm_id: "bbe9c69c24da4200bae0e4ecf6871de9"}
-- handling comm_msg
comm_msg Object {data: Object, comm_id: "bbe9c69c24da4200bae0e4ecf6871de9"}
-- handling comm_msg
comm_msg Object {data: Object, comm_id: "fa1d0a65945f465292719bf7a4a7c334"}
-- handling comm_msg
-- sync -- require of jupyter-js-widgets
-- sync -- require of jupyter-js-widgets
-- sync -- require of jupyter-js-widgets
-- sync -- require of jupyter-js-widgets

Failure case:

Starting WebSocket:
ws://192.168.99.100:9500/api/kernels/db8f9e6b-0d26-4dd5-a7ab-96b5308c669d
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "03e277ee19a84444894e3e28233d082d"}
-- async -- window.require of jupyter-js-widgets
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "0df5a65659ba4c9a919205aa42d09686"}
-- async -- window.require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 03e277ee19a84444894e3e28233d082d
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "3ddd27ee2cd54be19cd5a7ce58f83119"}
-- async -- window.require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 0df5a65659ba4c9a919205aa42d09686
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "096e0ab299ae4dcaad82b35f34ec3ba7"}
-- async -- window.require of jupyter-js-widgets
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "7cb420941ee24271a2e840eac496b2a9"}
-- async -- window.require of jupyter-js-widgets
new_model:  MODEL CREATED for comm_id 3ddd27ee2cd54be19cd5a7ce58f83119
comm_open Object {target_module: null, data: Object, target_name:
"jupyter.widget", comm_id: "f6c31741c06e4c5f9891ab2fb7c4c1a7"}
-- async -- window.require of jupyter-js-widgets
comm_msg Object {data: Object, comm_id: "7cb420941ee24271a2e840eac496b2a9"}
new_model:  MODEL CREATED for comm_id 096e0ab299ae4dcaad82b35f34ec3ba7
comm_msg Object {data: Object, comm_id: "7cb420941ee24271a2e840eac496b2a9"}
new_model:  MODEL CREATED for comm_id 7cb420941ee24271a2e840eac496b2a9
comm_msg Object {data: Object, comm_id: "f6c31741c06e4c5f9891ab2fb7c4c1a7"}
new_model:  MODEL CREATED for comm_id f6c31741c06e4c5f9891ab2fb7c4c1a7

Notice in the failure case that comm_msg arrive for some IDs before the
model has been created. Therefore, those messages aren't handled (notice no
instances of handling comm_msg).

We probably need to use promises from comm_open in order to chain incoming
comm_msg (for same ID). That way, the comm_open will handle finish being
handled (i.e. model created) before any comm_msgs are handled.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/ipython/ipywidgets/issues/561#issuecomment-220373688>

**Comment by jhpedemonte** [comment]
Actually, looks like `Kernel` in `jupyter-js-services` is trying to do promise chaining -- see `_commPromises`. I see a promise being stored in `_handleCommOpen` for a given ID, but then it is not found when querying in `_handleCommMsg`. Therefore, the comm msg is handled before model is created.

**Comment by jhpedemonte** [comment]
`CommManager.prototype.register_target` in the shim calls `ManagerBase.prototype.handle_comm_open` which creates the model async and returns a promise. However, `register_target` doesn't do anything with the promise. Neither does `Kernel.prototype._handleCommOpen`, which is what called `register_target`.  `_handleCommOpen`, though, immediately deletes the promise it had and later there is no promise chaining in `_handleCommMsg` (so msg is handled before model is created).

Not sure how to fix this.

**Comment by jhpedemonte** [comment]
Notebook does things a bit differently: https://github.com/jupyter/notebook/blob/master/notebook/static/services/kernels/comm.js#L78

Notice that the `response` is actually handled by this line:
```
return Promise.resolve(response).then(function() {return comm;});
```
ensuring that promises are chained here and in the `comm_msg` handler.

**Comment by jhpedemonte** [comment]
Did some surgery on the built index.js from web3. This seems to fix things for me:

```diff
--- index.built.js.orig	2016-05-19 12:03:28.000000000 -0500
+++ index.built.js	2016-05-19 12:01:48.000000000 -0500
@@ -37418,30 +37418,31 @@
 	 */
 	CommManager.prototype.register_target = function (target_name, f) {
 	    var handle = this.jsServicesKernel.registerCommTarget(target_name, function(jsServicesComm, msg) {
 	        // Create the comm.
 	        var comm = new Comm(jsServicesComm);
 
 	        // Call the callback for the comm.
 	        try {
 	            var response = f(comm, msg);
 	        } catch (e) {
 	            comm.close();
 	            var wrapped_error = new utils.WrappedError("Exception opening new comm", e);
 	            console.error(wrapped_error);
 	            return;
 	        }
+	        return response;
 	    });
 	    this.targets[target_name] = handle;
 	};
 
 	/**
 	 * Unregisters a comm target
 	 * @param  {string} target_name
 	 */
 	CommManager.prototype.unregister_target = function (target_name, f) {
 	    var handle = this.targets[target_name];
 	    handle.dispose();
 	    delete this.targets[target_name];
 	};
 
 
@@ -49964,33 +49965,35 @@
 	            console.error('Invalid comm message');
 	            return;
 	        }
 	        var content = msg.content;
 	        var targetName = content.target_name;
 	        var promise = utils.loadClass(content.target_name, content.target_module, this._targetRegistry).then(function (target) {
 	            var comm = new Comm(content.target_name, content.comm_id, _this._sendCommMessage.bind(_this), function () { _this._unregisterComm(content.comm_id); });
 	            try {
 	                var response = target(comm, msg);
 	            }
 	            catch (e) {
 	                comm.close();
 	                console.error("Exception opening new comm");
 	                throw (e);
 	            }
-	            _this._commPromises.delete(comm.commId);
-	            _this._comms.set(comm.commId, comm);
-	            return comm;
+	            return Promise.resolve(response).then(function() {
+		            _this._commPromises.delete(comm.commId);
+		            _this._comms.set(comm.commId, comm);
+		            return comm;
+		        });
 	        });
 	        this._commPromises.set(content.comm_id, promise);
 	    };
 	    /**
 	     * Handle 'comm_close' kernel message.
 	     */
 	    Kernel.prototype._handleCommClose = function (msg) {
 	        var _this = this;
 	        if (!validate.validateCommMessage(msg)) {
 	            console.error('Invalid comm message');
 	            return;
 	        }
 	        var content = msg.content;
 	        var promise = this._commPromises.get(content.comm_id);
 	        if (!promise) {
```

The first patch is to **services-shim.js**, to have it pass back the promise.  The 2nd part is to make the handlers in `jupyter-js-services` be more like those in the Notebook and actually take into account the `response`.

**Comment by jhpedemonte** [comment]
Created 2 PRs (see links above) which fix both scenarios I created ([one](https://github.com/jhpedemonte/ipywidgets/tree/web3-requirejs) & [two](https://github.com/jhpedemonte/ipywidgets/tree/requirejs-broken)).

**Comment by jhpedemonte** [comment]
These PRs also fix jupyter-incubator/dashboards_server#240.
